arangod/Aql/OptimizerRules.cpp

```diff
@@ -4625,36 +4632,51 @@ GeoIndexInfo geoDistanceFunctionArgCheck(
 
   if (setter1 != nullptr && setter2 != nullptr && setter1 == setter2 &&
       setter1->getType() == EN::ENUMERATE_COLLECTION) {
+
+    //get logical collection
     auto collNode = reinterpret_cast<EnumerateCollectionNode*>(setter1);
-    auto coll = collNode->collection();  // what kind of indexes does it have on
-                                         // what attributes
-    auto lcoll = coll->getCollection();
-    // TODO - check collection for suitable geo-indexes
-    for (auto indexShardPtr : lcoll->getIndexes()) {
+    auto coll = collNode->collection();
+    auto logicalColl = coll->getCollection();
+
+    //check for suitiable indexes
+    for (auto indexShardPtr : logicalColl->getIndexes()) {
       // get real index
       arangodb::Index& index = *indexShardPtr.get();
 
       // check if current index is a geo-index
-      if (index.type() != arangodb::Index::IndexType::TRI_IDX_TYPE_GEO1_INDEX &&
-          index.type() != arangodb::Index::IndexType::TRI_IDX_TYPE_GEO2_INDEX) {
+      std::size_t fieldNum = index.fields().size();
+      bool isGeo1 = (index.type() == arangodb::Index::IndexType::TRI_IDX_TYPE_GEO1_INDEX) && fieldNum == 1;
+      bool isGeo2 = (index.type() == arangodb::Index::IndexType::TRI_IDX_TYPE_GEO2_INDEX) && fieldNum == 2;
+
+      if (!isGeo1  && !isGeo2) {
         continue;
       }
 
-      TRI_ASSERT(index.fields().size() == 2);
-
-      // check access paths of attributes in ast and those in index match
-      if (index.fields()[0] == attributeAccess1.second &&
-          index.fields()[1] == attributeAccess2.second) {
-        info.collectionNode = collNode;
-        info.index = indexShardPtr;
-        TRI_AttributeNamesJoinNested(attributeAccess1.second, info.longitude,
-                                     true);
-        TRI_AttributeNamesJoinNested(attributeAccess2.second, info.latitude,
-                                     true);
-        return info;
-      }
-    }
-  }
+      if(isGeo2){
+        // check access paths of attributes in ast and those in index match
+        if (index.fields()[0] == attributeAccess1.second && index.fields()[1] == attributeAccess2.second) {
+          info.collectionNode = collNode;
+          info.index = indexShardPtr;
+          TRI_AttributeNamesJoinNested(attributeAccess1.second, info.longitude, true);
+          TRI_AttributeNamesJoinNested(attributeAccess2.second, info.latitude, true);
+          return info;
+        }
+      } else if (isGeo1){
+        std::vector<basics::AttributeName> fields1 = index.fields()[0];
+        std::vector<basics::AttributeName> fields2 = index.fields()[0];
+        fields1.back().name += "[0]";
```

Comment:
looks just wrong. shouldn'it be `index.fields()[1]` here?

rethinking: the index has just a single field, so the `[0]` is probably actually correct here.
A comment would have clarified that for the quick reader.